### PR TITLE
feat: support a color or colour variable from tool change macros

### PR DESCRIPTION
**Pull request by richardjm**
When set the button color will use that set from the variable.

image

Example macros

```
############################################################################
# Complementary functions controlling variables used by mainsail for display
############################################################################

# Requires appropriately named variables in your tool change macros (these are used by mainsail)
# [gcode_macro T0]
# variable_active: 0
# variable_color: "undefined"

[gcode_macro _SET_ACTIVE_TOOL]
description: Sets the active tool in mainsail (and all others inactive)
gcode:
  {% set TOOL = params.TOOL|default(-1)|int %}
  {% for T in range(9) %}
    {% if T == TOOL %}
      SET_GCODE_VARIABLE MACRO=T{T} VARIABLE=active VALUE=1
    {% else %}
      SET_GCODE_VARIABLE MACRO=T{T} VARIABLE=active VALUE=0
    {% endif %}
  {% endfor %}

# _SET_TOOL_COLOR TOOL=<id> COLOR=<color>
# e.g. _SET_TOOL_COLOR TOOL=0 COLOR=#112233
[gcode_macro _SET_TOOL_COLOR]
description: Sets a color for mainsail extruder
gcode:
  {% set tool = params.TOOL|default(-1)|int %}
  # get color from rawparams and and remove the # char from the color code if present
  {% set ns = namespace(color="undefined") %}
  {% for param in rawparams.split(' ') %}
    {% if 'COLOR' in param|string|upper %}
      {% set rawcolor = (param.split('='))[1] %}
      {% set ns.color = rawcolor if rawcolor[0]|lower in ['0','1','2','3','4','5','6','7','8','9','a','b','c','d','e','f'] else rawcolor[1:] %}
    {% endif %}
  {% endfor %}
  SET_GCODE_VARIABLE MACRO=T{tool} VARIABLE=color VALUE='"{ns.color}"'
  # Requires [save_variables], for me this is provided by ercf
  {% if printer.save_variables is defined %}
    SAVE_VARIABLE VARIABLE=t{tool}_color VALUE='"{ns.color}"'
  {% endif %}

# Reload the colors from the [save_variables] if used
[delayed_gcode _SET_TOOL_COLORS_ON_STARTUP]
initial_duration: 1
gcode:
  {% set svv = printer.save_variables.variables %}
  {% for T in range(9) %}
    _SET_TOOL_COLOR TOOL={T} COLOR={svv['t' + (T|string) + '_color']}
  {% endfor %}
```

Signed-off-by: Richard Mitchell <richardjm+mainsail@gmail.com>

### CODE REVIEW

**Review by meteyou — CHANGES_REQUESTED** [review]

File: src/store/printer/getters.ts [hunk 1]
```diff
@@ -1097,6 +1097,7 @@ export const getters: GetterTree<PrinterState, RootState> = {
                 tools.push({
                     name: macro.name,
                     active: macro.variables.active ?? false,
+                    color: macro.variables.color ?? macro.variables.colour ?? 'null',
```
[the remark this answers is not in the document]
Comment (richardjm) [reply src/store/printer/getters.ts:1100]: Yes was just trying to support both spellings of colour.

**Comment by zellneralex** [comment]
```[gcode_macro _SET_TOOL_COLOR]
description: Sets a color for mainsail extruder (e.g. TOOL=1 COLOR=FF00FF)
gcode:
  {% set tool = params.TOOL|default(-1)|int %}
  # get color from rawparams and remove #
  {% set ns = namespace(color="undefined") %}
  {% for param in rawparams.split(' ') %}
    {% if 'COLOR' in param|string %}
      {% set ns.color = param[7:] %}
    {% endif %}
  {% endfor %}
  SET_GCODE_VARIABLE MACRO=T{tool} VARIABLE=color VALUE='"{ns.color}"'
  # Requires [save_variables], for me this is provided by ercf
  {% if printer.save_variables.variables is defined %}
    SAVE_VARIABLE VARIABLE=t{tool}_color VALUE='"{ns.color}"'
  {% endif %}``` avoids the need of a post processing of the gcode

**Comment by zellneralex** [comment]
That version will handle both correct so you can use #12345 or 12345 as COLOR 
```
[gcode_macro _SET_TOOL_COLOR]
description: Sets a color for mainsail extruder
gcode:
  {% set tool = params.TOOL|default(-1)|int %}
  # get color from rawparams and remove #
  {% set ns = namespace(color="undefined") %}
  #{% set rawparams_list = rawparams.split(' ') %}
  {% for param in rawparams.split(' ') %}
    {% if 'COLOR' in param|string %}
      {% set rawcolor = (param.split('='))[1] %}
      {% set ns.color = rawcolor if rawcolor[0]|lower in ['0','1','2','3','4','5','6','7','8','9','a','b','c','d','e','f'] else rawcolor[1:] %}
    {% endif %}
  {% endfor %}
  SET_GCODE_VARIABLE MACRO=T{tool} VARIABLE=color VALUE='"{ns.color}"'
  # Requires [save_variables], for me this is provided by ercf
  {% if printer.save_variables.variables is defined %}
    SAVE_VARIABLE VARIABLE=t{tool}_color VALUE='"{ns.color}"'
  {% endif %}
```

File: src/components/panels/ExtruderControlPanel.vue [hunk 2]
```diff
@@ -63,6 +63,7 @@
                             :key="tool.name"
                             :class="tool.active ? 'primary--text' : {}"
                             :disabled="isPrinting"
+                            :style="'background-color: #' + tool.color + ';'"
```
[the remark this answers is not in the document]
Comment (richardjm) [reply src/components/panels/ExtruderControlPanel.vue:66]: This is what already happens as the default value for tool.color is null via
```
color: macro.variables.colour ?? null,
```

[the remark this answers is not in the document]
Comment (meteyou) [reply src/components/panels/ExtruderControlPanel.vue:66]: But even if the color is null, `background-color: #;` is executed

[the remark this answers is not in the document]
Comment (richardjm) [reply src/components/panels/ExtruderControlPanel.vue:66]: Not as far as I can see as the null coalesces into null removing the background-color. Can't find the exact documentation for it but as you can see in the screenshot style is completely removed.

**Comment by richardjm** [comment]
Slight modification to the macro as it wasn't working with `_SET_TOOL_COLOR tool=1 color=FF0000` (lower case COLOR argument)

```
# _SET_TOOL_COLOR TOOL=<id> COLOR=<color>
# e.g. _SET_TOOL_COLOR TOOL=0 COLOR=#112233
[gcode_macro _SET_TOOL_COLOR]
description: Sets a color for mainsail extruder
gcode:
  {% set tool = params.TOOL|default(-1)|int %}
  # get color from rawparams and and remove the # char from the color code if present
  {% set ns = namespace(color="undefined") %}
  {% for param in rawparams.split(' ') %}
    {% if 'COLOR' in param|string|upper %}
      {% set rawcolor = (param.split('='))[1] %}
      {% set ns.color = rawcolor if rawcolor[0]|lower in ['0','1','2','3','4','5','6','7','8','9','a','b','c','d','e','f'] else rawcolor[1:] %}
    {% endif %}
  {% endfor %}
  SET_GCODE_VARIABLE MACRO=T{tool} VARIABLE=color VALUE='"{ns.color}"'
  # Requires [save_variables], for me this is provided by ercf
  {% if printer.save_variables is defined %}
    SAVE_VARIABLE VARIABLE=t{tool}_color VALUE='"{ns.color}"'
  {% endif %}
  ```

File: src/components/panels/ExtruderControlPanel.vue (hunk 2 above)
[the remark this answers is not in the document]
Comment (richardjm) [reply src/components/panels/ExtruderControlPanel.vue]: Modified to use object setting which I believe I have confirmed now - https://shorturl.at/rO347

File: src/store/printer/getters.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (richardjm) [reply src/store/printer/getters.ts]: Changed to use `color` spelling consistently

**Comment by zellneralex** [comment]
I promise that will be the last iteration. Now we also insure that only COLOR will work as parameter and not anything that includes COLOR in the parameter name
```
# _SET_TOOL_COLOR TOOL=<id> COLOR=<color>
# e.g. _SET_TOOL_COLOR TOOL=0 COLOR=#112233
[gcode_macro _SET_TOOL_COLOR]
description: Sets a color for mainsail extruder
gcode:
  {% set tool = params.TOOL|default(-1)|int %}
  # get color from rawparams and and remove the # char from the color code if present
  {% set ns = namespace(color="undefined") %}
  {% for param in rawparams.split(' ') %}
    {% if (param.split('='))[0]|lower == 'color' %}
      {% set rawcolor = (param.split('='))[1] %}
      {% set ns.color = rawcolor if rawcolor[0]|lower in ['0','1','2','3','4','5','6','7','8','9','a','b','c','d','e','f'] else rawcolor[1:] %}
    {% endif %}
  {% endfor %}
  {action_respond_info(ns.color)}
  SET_GCODE_VARIABLE MACRO=T{tool} VARIABLE=color VALUE='"{ns.color}"'
  # save value to file if [save_variables] is defined in printer.cfg
  {% if printer.save_variables is defined %}
    SAVE_VARIABLE VARIABLE=t{tool}_color VALUE='"{ns.color}"'
  {% endif %}
  ```

**Comment by zellneralex** [comment]
Ok the reason for the change can be found at the mainsail discord. Will try to add the picture later.

```

# _SET_TOOL_COLOR TOOL=<id> COLOR=<color>
# e.g. _SET_TOOL_COLOR TOOL=0 COLOR=#112233
[gcode_macro _SET_TOOL_COLOR]
description: Sets a color for mainsail extruder
gcode:
  # get color and tool from rawparams and and remove the # char from the color code if present
  {% set ns = namespace(color="undefined", tool=-1) %}
  {% for param in rawparams.split(' ') %}
    {% if (param.split('='))[0]|lower == 'color' %}
      {% set rawcolor = (param.split('='))[1] %}
      {% set ns.color = rawcolor if rawcolor[0]|lower in ['0','1','2','3','4','5','6','7','8','9','a','b','c','d','e','f'] else rawcolor[1:] %}
    {% endif %}
    {% if (param.split('='))[0]|lower == 'tool' %}
      {% set ns.tool = (param.split('='))[1]|int %}
    {% endif %}
  {% endfor %}
  # set color if all is ok or error out
  {% if ns.tool == -1 %} 
    {action_raise_error("TOOL parameter missing")}
  {% elif printer['gcode_macro T' + ns.tool|string] is not defined %}
    {action_raise_error("No macro T%d defined" % (ns.tool))}
  {% elif printer['gcode_macro T' + ns.tool|string].color is not defined %}
    {action_raise_error("variable_color missing at macro T%d" % (ns.tool))}
  {% else %}
    SET_GCODE_VARIABLE MACRO=T{ns.tool} VARIABLE=color VALUE='"{ns.color}"'
    # save value to file if [save_variables] is defined in printer.cfg
    {% if printer.save_variables is defined %}
      SAVE_VARIABLE VARIABLE=t{ns.tool}_color VALUE='"{ns.color}"'
    {% endif %}
  {% endif %}
```

File: src/components/panels/ExtruderControlPanel.vue (hunk 2 above)
[the remark this answers is not in the document]
Comment (richardjm) [reply src/components/panels/ExtruderControlPanel.vue]: Wasn't to check the new code?

File: src/store/printer/getters.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (meteyou) [reply src/store/printer/getters.ts]: ohhh... i overlooked the `colour` <> `color` difference

**Review by meteyou — APPROVED** [review]

File: src/store/printer/getters.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (richardjm) [reply src/store/printer/getters.ts]: I've just re-instated that. Anything else I need to do in order to get this merged?

**Pull request merged**